### PR TITLE
machine refactor 3: add symlinks for sockets

### PR DESCRIPTION
**Pull request by baude**
to avoid errors on macos, we use symlinks to long socket names.

Fixes: #12751
Fixes: #13609

Signed-off-by: Brent Baude <bbaude@redhat.com>

[NO NEW TESTS NEEDED]

Signed-off-by: Brent Baude <bbaude@redhat.com>

### CODE REVIEW

File: pkg/machine/qemu/options_linux.go [hunk 1]
```diff
@@ -11,3 +11,8 @@ func getRuntimeDir() (string, error) {
 	}
 	return util.GetRuntimeDir()
 }
+
+// makeSymlink is a no-op for Linux
+func (m *MachineFile) makeSymlink(*string) error {
```
Comment (Luap99) [inline pkg/machine/qemu/options_linux.go:16]: Why is this a noop for linux, we have the same problem there just that the limit is 108 vs 104 on darwin.

File: pkg/machine/qemu/options_darwin.go [hunk 2]
```diff
@@ -11,3 +12,15 @@ func getRuntimeDir() (string, error) {
 	}
 	return tmpDir, nil
 }
+
+// makeSymlink for macOS creates a symlink in $HOME/.podman/
+// for a machinefile like a socket
+func (m *MachineFile) makeSymlink(symlink *string) error {
+	homedir, err := os.UserHomeDir()
```
Comment (Luap99) [inline pkg/machine/qemu/options_darwin.go:19]: before doing this would it make sense to check if path > 104 and skip it otherwise?

**Comment by ashley-cui** [comment]
```
testuserwithreallylongusernamehahahahahahaha@Ashleys-Mini darwin % ./podman machine startStarting machine "podman-machine-default"INFO[0000] waiting for clients...                       ERRO[0000] Error listening on socket: /Users/testuserwithreallylongusernamehahahahahahaha/.local/share/containers/podman/machine/podman-machine-default/podman.sock: listen unix /Users/testuserwithreallylongusernamehahahahahahaha/.local/share/containers/podman/machine/podman-machine-default/podman.sock: bind: invalid argument Error: dial unix /var/folders/2j/8l8d892j0fv0rjz_mf0y3y5c0000gp/T/podman/qmp_podman-machine-default.sock: connect: no such file or directory
slightlyshorterusername@Ashleys-Mini darwin % ./podman machine startStarting machine "podman-machine-default"INFO[0000] waiting for clients...                       ERRO[0000] Error listening on socket: /Users/slightlyshorterusername/.local/share/containers/podman/machine/podman-machine-default/podman.sock: listen unix /Users/slightlyshorterusername/.local/share/containers/podman/machine/podman-machine-default/podman.sock: bind: invalid argument Error: dial unix /var/folders/y0/cq7kq5kd29s0yshf5qw0k2d40000gq/T/podman/qmp_podman-machine-default.sock: connect: no such file or directory
reasonable@Ashleys-Mini darwin % ./podman machine startStarting machine "podman-machine-default"INFO[0000] waiting for clients...                       INFO[0000] new connection from  to /var/folders/pq/tjzq5b_17jx35fjdkk8_6t6h0000gr/T/podman/qmp_podman-machine-default.sock Waiting for VM ...INFO[0027] Socket forward established: /Users/reasonable/.local/share/containers/podman/machine/podman-machine-default/podman.sock to /run/user/504/podman/podman.sock ...Machine "podman-machine-default" started successfully
```

**Comment by baude** [comment]
@containers/podman-maintainers PTAL

File: pkg/machine/qemu/machine.go [hunk 3]
```diff
@@ -1116,43 +1116,44 @@ func (v *MachineVM) setupAPIForwarding(cmd []string) ([]string, string, apiForwa
 		forwardUser = "root"
 	}
 
-	cmd = append(cmd, []string{"-forward-sock", socket}...)
+	cmd = append(cmd, []string{"-forward-sock", socket.GetPath()}...)
 	cmd = append(cmd, []string{"-forward-dest", destSock}...)
 	cmd = append(cmd, []string{"-forward-user", forwardUser}...)
 	cmd = append(cmd, []string{"-forward-identity", v.IdentityPath}...)
-	link := filepath.Join(filepath.Dir(filepath.Dir(socket)), "podman.sock")
+	//link := filepath.Join(filepath.Dir(filepath.Dir(socket)), "podman.sock")
```
Comment (ashley-cui) [inline pkg/machine/qemu/machine.go:1123]: Dev artifact?

**Review by ashley-cui — COMMENTED** [review]
Works for me now! 
```
testuserwithreallylongusernamehahahahahahaha@Ashleys-Mini darwin % ./podman machine start hellohellohellohellohellohello
Starting machine "hellohellohellohellohellohello"
INFO[0000] waiting for clients...                       
INFO[0000] new connection from  to /tmp/podman/qmp_hellohellohellohellohellohello.sock 
Waiting for VM ...
INFO[0015] Socket forward established: /Users/testuserwithreallylongusernamehahahahahahaha/.podman/podman.sock to /run/user/502/podman/podman.sock 
...
Machine "hellohellohellohellohellohello" started successfully
```

Other than a small dev artifact, LGTM!

**Comment by TomSweeneyRedHat** [comment]
LGTM

**Review by vrothberg — APPROVED** [review]
/lgtm

**Pull request merged**